// testcafe-reporter-dashboard/lib/types/dashboard.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var AggregateCommandType;
(function (AggregateCommandType) {
    AggregateCommandType["reportTaskStart"] = "reportTaskStart";
    AggregateCommandType["reportTestStart"] = "reportTestStart";
    AggregateCommandType["reportTestDone"] = "reportTestDone";
    AggregateCommandType["reportTaskDone"] = "reportTaskDone";
    AggregateCommandType["createUpload"] = "createUpload";
})(AggregateCommandType = exports.AggregateCommandType || (exports.AggregateCommandType = {}));
;
var AggregateNames;
(function (AggregateNames) {
    AggregateNames["Run"] = "Run";
    AggregateNames["Upload"] = "Upload";
})(AggregateNames = exports.AggregateNames || (exports.AggregateNames = {}));
;
exports.createTestError = (error, errorModel) => ({
    code: error.code,
    testRunPhase: error.testRunPhase,
    errorModel: errorModel
});
exports.createDashboardTestRunInfo = (testRunInfo, browserRuns) => ({
    quarantine: testRunInfo.quarantine,
    unstable: testRunInfo.unstable,
    warnings: testRunInfo.warnings,
    browserRuns: browserRuns
});
var UploadStatus;
(function (UploadStatus) {
    UploadStatus["Completed"] = "Completed";
    UploadStatus["Failed"] = "Failed";
})(UploadStatus = exports.UploadStatus || (exports.UploadStatus = {}));
;
;
