// testcafe-reporter-dashboard/lib/report-commands-factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const dashboard_1 = require("./types/dashboard");
function reportCommandsFactory(reportId, transport) {
    async function sendReportCommand(type, payload) {
        return transport.sendResolveCommand({
            aggregateId: reportId,
            aggregateName: dashboard_1.AggregateNames.Run,
            type,
            payload
        });
    }
    return {
        sendTaskStartCommand(payload) {
            return sendReportCommand(dashboard_1.AggregateCommandType.reportTaskStart, payload);
        },
        sendTestStartCommand(payload) {
            return sendReportCommand(dashboard_1.AggregateCommandType.reportTestStart, payload);
        },
        sendTestDoneCommand(payload) {
            return sendReportCommand(dashboard_1.AggregateCommandType.reportTestDone, payload);
        },
        sendTaskDoneCommand(payload) {
            return sendReportCommand(dashboard_1.AggregateCommandType.reportTaskDone, payload);
        }
    };
}
exports.default = reportCommandsFactory;
;
